packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -0,0 +1,362 @@
+        `  /opt/aws/bin/cfn-signal -e ${errCode} ${resourceLocator}`,
+        ...printLog ? ['  cat /var/log/cfn-init.log >&2'] : [],
+        ')',
+      ]);
+    }
+  }
+
+  private bind(scope: Construct, options: AttachInitOptions): { configData: any, authData: any } {
+    const nonEmptyConfigs = mapValues(this._configs, c => c.isEmpty ? undefined : c);
+
+    const configNameToBindResult = mapValues(nonEmptyConfigs, c => c._bind(scope, options));
+
+    return {
+      configData: {
+        configSets: mapValues(this._configSets, configNames => configNames.filter(name => nonEmptyConfigs[name] !== undefined)),
+        ...mapValues(configNameToBindResult, c => c.config),
+      },
+      authData: Object.values(configNameToBindResult).map(c => c.authentication).reduce(deepMerge, undefined),
+    };
+  }
+
+}
+
+/**
+ * Options for attach a CloudFormationInit to a resource
+ */
+export interface AttachInitOptions {
+  /**
+   * Instance role of the consuming instance or fleet
+   */
+  readonly instanceRole: iam.IRole;
+
+  /**
+   * OS Platfrom the init config will be used for
+   */
+  readonly platform: InitPlatform;
+
+  /**
+   * UserData to add commands to
+   */
+  readonly userData: UserData;
+
+  /**
+   * ConfigSet to activate
+   *
+   * @default ['default']
+   */
+  readonly configSets?: string[];
+
+  /**
+   * Whether to embed a hash into the userData
+   *
+   * If `true` (the default), a hash of the config will be embedded into the
+   * UserData, so that if the config changes, the UserData changes and
+   * the instance will be replaced.
+   *
+   * If `false`, no such hash will be embedded, and if the CloudFormation Init
+   * config changes nothing will happen to the running instance.
+   *
+   * @default true
+   */
+  readonly embedFingerprint?: boolean;
+
+  /**
+   * Print the results of running cfn-init to the Instance System Log
+   *
+   * By default, the output of running cfn-init is written to a log file
+   * on the instance. Set this to `true` to print it to the System Log
+   * (visible from the EC2 Console), `false` to not print it.
+   *
+   * (Be aware that the system log is refreshed at certain points in
+   * time of the instance life cycle, and successful execution may
+   * not always show up).
+   *
+   * @default true
+   */
+  readonly printLog?: boolean;
+
+  /**
+   * Don't fail the instance creation when cfn-init fails
+   *
+   * You can use this to prevent CloudFormation from rolling back when
+   * instances fail to start up, to help in debugging.
+   *
+   * @default false
+   */
+  readonly ignoreFailures?: boolean;
+}
+
+/**
+ * A collection of configuration elements
+ */
+export class InitConfig {
+  private readonly elements = new Array<InitElement>();
+
+  constructor(elements: InitElement[]) {
+    this.add(...elements);
+  }
+
+  /**
+   * Whether this configset has elements or not
+   */
+  public get isEmpty() {
```

Comment:
getter here is weird. Prefer to have this as  a method instead? 

`.isEmpty()` feels more natural than `.isEmpty`